Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -239,21 +250,257 @@ FieldValue Serializer::DecodeFieldValue(const uint8_t* bytes, size_t length) {
+  // Implementation note: This is roughly modeled on pb_encode_delimited (which
+  // is actually pb_encode_submessage), adjusted to account for the oneof in
+  // FieldValue.
+
+  // First calculate the message size using a non-writing substream.
+  pb_ostream_t substream = PB_OSTREAM_SIZING;
+  EncodeFieldValueImpl(&substream, field_value);
+  size_t size = substream.bytes_written;
+
+  // Write out the size to the output stream.
+  EncodeVarint(stream, size);
+
+  // If stream is itself a sizing stream, then we don't need to actually parse
+  // field_value a second time; just update the bytes_written via a call to
+  // pb_write. (If we try to write the contents into a sizing stream, it'll
+  // fail since sizing streams don't actually have any buffer space.)
+  if (stream->callback == NULL) {
+    bool status = pb_write(stream, NULL, size);
+    if (!status) {
+      // TODO(rsgowman): figure out error handling
+      abort();
+    }
+    return;
+  }
+
+  // Ensure the output stream has enough space
+  if (stream->bytes_written + size > stream->max_size) {
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+
+  // Use a substream to verify that a callback doesn't write more than what it
+  // did the first time. (Use an initializer rather than setting fields
+  // individually like nanopb does. This gives us a *chance* of noticing if
+  // nanopb adds new fields.)
+  substream = {stream->callback, stream->state, /*max_size=*/size,
+               /*bytes_written=*/0, /*errmsg=*/NULL};
+
+  EncodeFieldValueImpl(&substream, field_value);
+  stream->bytes_written += substream.bytes_written;
+  stream->state = substream.state;
+  stream->errmsg = substream.errmsg;
+
+  if (substream.bytes_written != size) {
+    // submsg size changed
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+}
+
+FieldValue DecodeSubFieldValue(pb_istream_t* stream) {
+  // Implementation note: This is roughly modeled on pb_decode_delimited,
+  // adjusted to account for the oneof in FieldValue.
+  pb_istream_t substream;
+  bool status = pb_make_string_substream(stream, &substream);
+  if (!status) {
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+
+  FieldValue fv = DecodeFieldValueImpl(&substream);
+
+  // NB: future versions of nanopb read the remaining characters out of the
+  // substream (and return false if that fails) as an additional safety
+  // check within pb_clost_string_substream. Unfortunately, that's not present
+  // in the current version (0.38).  We'll make a stronger assertion and check
+  // to make sure there *are* no remaining characters in the substream.
+  if (substream.bytes_left != 0) {
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+  pb_close_string_substream(stream, &substream);
+
+  return fv;
+}
+
+void EncodeFieldEntry(
+    pb_ostream_t* stream,
+    const std::pair<const std::string, const FieldValue>& kv) {
```

Comment:
I think it would be clearer to use `value_type` here, which should be possible with an alias:
```cpp
const ObjectT::value_type& kv
```

Reply:
deferred to chat